invenio_drafts_resources/records/models.py

```diff
@@ -17,24 +17,109 @@
 
 
-def ParentRecordMixin(parent_record_cls):
+class ParentRecordMixin:
     """A mixin factory that add the foreign keys to the parent record.
 
-    It is intended to be added to the "child" record class, e.g.:
-    ``class MyRecord(RecordBase, ParentRecordMixin(MyRecordParentClass))``.
+    Usage:
+
+    .. code-block:: python
+
+        class MyRecord(db.Model, RecordMetadataBase, ParentRecordMixin):
+            __parent_record_model__ = ParentRecordMetadata
+    """
+
+    __parent_record_model__ = None
+
+    @declared_attr
+    def parent_id(cls):
+        """Parent identifier."""
+        # We restrict deletion of the parent record in case a record or draft
+        # exists via database-level trigger.
+        return db.Column(UUIDType, db.ForeignKey(
+            cls.__parent_record_model__.id, ondelete="RESTRICT"))
+
+    @declared_attr
+    def parent(cls):
+        """Relationship to parent record."""
+        return db.relationship(cls.__parent_record_model__)
+
+    parent_index = db.Column(db.Integer, nullable=True)
+    """The version index of the record."""
+
+
+class ParentRecordStateMixin:
+    """Database model mixin to keep the state of the latest and next version.
+
+    We keep this data outside the parent record itself, because we want to
+    update it without impacting the parent record's version counter. The
+    version counter in the parent record we use to determine if we have to
+    reindex all record versions.
+
+    Usage:
+
+    .. code-block:: python
+
+        class MyParentState(db.Model, ParentRecordState):
+            parent_record_model = MyParentRecord
+            draft_model = MyDraft
+            record_model = MyRecord
+    """
+
+    __parent_record_model__ = None
+    __record_model__ = None
+    __draft_model__ = None
+
+    @declared_attr
+    def parent_id(cls):
+        """Parent record identifier."""
+        return db.Column(
+            UUIDType,
+            # 1) If the parent record is deleted, we automatically delete
+            # the state as well via database-level on delete trigger.
+            # 2) A parent can only be deleted, if it has no drafts/records via
+            # FKs from drafts/records to the parent.
+            db.ForeignKey(cls.__parent_record_model__.id, ondelete="CASCADE"),
+            primary_key=True,
+        )
+
+    @declared_attr
+    def latest_id(cls):
+        """UUID of the latest published record/draft.
+
+        Note, since a record and draft share the same UUID, the UUID can be
+        used to get both the record or the draft. It's a foreign key to the
+        record to ensure that the record exists (and thus is published).
+
+        If no record has been published, the value is None.
+        """
+        return db.Column(
+            UUIDType,
+            db.ForeignKey(cls.__record_model__.id),
+            nullable=True,
+        )
+
+    latest_index = db.Column(db.Integer, nullable=True)
+    """The index of the latest published record.
```

Comment:
The reasoning for both ``latest_id`` and ``latest_index`` is that we want to avoid expensive select queries. Both properties could be queried by a select similar to:

```sql
SELECT id as latest_id, parent_index as latest_index FROM rdm_records_metadata WHERE parent_id=:id ORDER BY parent_index DESC LIMIT 1
```

Instead, we can do a single primary key look like this:

```sql
SELECT latest_id, latest_index FROM rdm_versions_state WHERE parent_id=:id
```

It's a classic search problem - you can either have 1) fast inserts and expensive selects OR 2) expensive inserts and fast selects :-)

Reply:
Got it: in our case we are making the insert a bit more involved (need to update `latest_index` / `latest_id` in this table in addition to the regular `rdm_records_metadata` table), but we get a faster select (which is used in insertion ;) )